File: func8.js
// func8

// Semigroup: An associative type with a concat method. 
// Monoid: A Semigroup with a neutral element (empty method)

const Sum = x => ({
    x,
    concat: ({x: y}) => Sum(x + y),
    inspect: () => `Sum(${x})`
})

Sum.empty = () => Sum(0)

// const result = Sum.empty().concat(Sum(1).concat(Sum(2))) // want Sum(3)

const All = x => ({
    x,
    concat: ({x: y}) => All(x && y),
    inspect: () => `All(${x})`
})

All.empty = () => All(true)

// const result = All(true).concat(All(false)).concat(All.empty()) // All(false)


const First = x => ({
    x,
    concat: () => First(x),
    inspect: () => `First(${x})`
})

// There is no way to define a neutral element on First, thus First remains a Semi-group and cannot be a Monoid.

// const result = First("blah").concat(First("ice cream")) // First(false)

const sum = xs =>
    xs.reduce((acc, x) => acc + x, 0)

const result = sum([1,2,3])


console.log(result)

File: func8.py
# func8

# Monoid: A Semigroup with a neutral element

class Sum:
    def __init__(self, x):
        self.x = x
        self.concat = lambda o: Sum(x + o.x)
    
    def __repr__(self):
        return 'Sum({})'.format(self.x)
    
    @staticmethod
    def empty():
        return Sum(0)


# result = Sum.empty().concat(Sum(1).concat(Sum(2)))


class All:
    def __init__(self, x):
        self.x = x
        self.concat = lambda o: All(x & o.x)
    
    def __repr__(self):
        return 'All({})'.format(self.x)

    @staticmethod
    def empty():
        return All(True)


result = All(True).concat(All(False)).concat(All.empty())

class First:
    def __init__(self, x):
        self.x = x
        self.concat = lambda o: First(x)
    
    def __repr__(self):
        return 'First({})'.format(self.x)


# result = First('blah').concat(First('ice cream'))

print(result)